File: komputer-sdk/typescript/src/integration.test.ts
/**
 * Integration tests for the KomputerClient.
 *
 * These tests require a running komputer.ai API.
 * Set KOMPUTER_API_URL to point to the server (default: http://localhost:8080).
 *
 * Run with:
 *   npx vitest run src/integration.test.ts
 */

import { describe, it, expect, beforeAll, afterAll } from "vitest";
import { KomputerClient } from "./client";

declare const process: { env: Record<string, string | undefined> };
const BASE_URL: string = process.env.KOMPUTER_API_URL || "http://localhost:8080";
const client = new KomputerClient(BASE_URL);

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

async function tryDelete(fn: () => Promise<unknown>): Promise<void> {
  try {
    await fn();
  } catch {
    // ignore — resource may not exist
  }
}

// ---------------------------------------------------------------------------
// Memories CRUD
// ---------------------------------------------------------------------------

describe("Memories CRUD", () => {
  const name = "sdk-test-ts-memory-crud";

  beforeAll(async () => {
    await tryDelete(() => client.deleteMemory(name));
  });

  afterAll(async () => {
    await tryDelete(() => client.deleteMemory(name));
  });

  it("creates a memory", async () => {
    const mem = await client.createMemory({
      name,
      content: "initial content",
      description: "integration test memory",
    });
    expect(mem).toBeDefined();
    expect((mem as any).name).toBe(name);
  });

  it("gets the created memory", async () => {
    const mem = await client.getMemory(name);
    expect(mem).toBeDefined();
    expect((mem as any).name).toBe(name);
  });

  it("lists memories and finds the created one", async () => {
    const list = await client.listMemories();
    const items: any[] = Array.isArray(list) ? list : (list as any).memories ?? (list as any).items ?? [];
    const found = items.some((m: any) => m.name === name);
    expect(found).toBe(true);
  });

  it("patches the memory", async () => {
    const patched = await client.patchMemory({
      name,
      content: "patched content",
    });
    expect(patched).toBeDefined();
  });

  it("get reflects the patched content", async () => {
    const mem = await client.getMemory(name);
    const content: string = (mem as any).content ?? (mem as any).spec?.content ?? "";
    expect(content).toContain("patched");
  });

  it("deletes the memory", async () => {
    await expect(client.deleteMemory(name)).resolves.not.toThrow();
  });

  it("get throws after delete", async () => {
    await expect(client.getMemory(name)).rejects.toThrow();
  });
});

// ---------------------------------------------------------------------------
// Memories idempotent create
// ---------------------------------------------------------------------------

describe("Memories idempotent create", () => {
  const name = "sdk-test-ts-memory-idempotent";

  beforeAll(async () => {
    await tryDelete(() => client.deleteMemory(name));
  });

  afterAll(async () => {
    await tryDelete(() => client.deleteMemory(name));
  });

  it("creates a memory the first time", async () => {
    await expect(
      client.createMemory({ name, content: "first content" })
    ).resolves.not.toThrow();
  });

  it("creates the same memory again with different content without error", async () => {
    await expect(
      client.createMemory({ name, content: "second content" })
    ).resolves.not.toThrow();
  });

  it("get reflects the updated content", async () => {
    const mem = await client.getMemory(name);
    const content: string = (mem as any).content ?? (mem as any).spec?.content ?? "";
    expect(content).toContain("second");
  });
});

// ---------------------------------------------------------------------------
// Skills CRUD
// ---------------------------------------------------------------------------

describe("Skills CRUD", () => {
  const name = "sdk-test-ts-skill-crud";

  beforeAll(async () => {
    await tryDelete(() => client.deleteSkill(name));
  });

  afterAll(async () => {
    await tryDelete(() => client.deleteSkill(name));
  });

  it("creates a skill", async () => {
    const skill = await client.createSkill({
      name,
      content: "echo hello",
      description: "integration test skill",
    });
    expect(skill).toBeDefined();
    expect((skill as any).name).toBe(name);
  });

  it("gets the created skill", async () => {
    const skill = await client.getSkill(name);
    expect(skill).toBeDefined();
    expect((skill as any).name).toBe(name);
  });

  it("lists skills and finds the created one", async () => {
    const list = await client.listSkills();
    const items: any[] = Array.isArray(list) ? list : (list as any).skills ?? (list as any).items ?? [];
    const found = items.some((s: any) => s.name === name);
    expect(found).toBe(true);
  });

  it("patches the skill", async () => {
    const patched = await client.patchSkill({
      name,
      content: "echo patched",
      description: "patched description",
    });
    expect(patched).toBeDefined();
  });

  it("get reflects the patched content", async () => {
    const skill = await client.getSkill(name);
    const content: string = (skill as any).content ?? (skill as any).spec?.content ?? "";
    expect(content).toContain("patched");
  });

  it("deletes the skill", async () => {
    await expect(client.deleteSkill(name)).resolves.not.toThrow();
  });

  it("get throws after delete", async () => {
    await expect(client.getSkill(name)).rejects.toThrow();
  });
});

// ---------------------------------------------------------------------------
// Skills idempotent create
// ---------------------------------------------------------------------------

describe("Skills idempotent create", () => {
  const name = "sdk-test-ts-skill-idempotent";

  beforeAll(async () => {
    await tryDelete(() => client.deleteSkill(name));
  });

  afterAll(async () => {
    await tryDelete(() => client.deleteSkill(name));
  });

  it("creates a skill the first time", async () => {
    await expect(
      client.createSkill({ name, content: "echo first", description: "first" })
    ).resolves.not.toThrow();
  });

  it("creates the same skill again with different content without error", async () => {
    await expect(
      client.createSkill({ name, content: "echo second", description: "second" })
    ).resolves.not.toThrow();
  });

  it("get reflects the updated content", async () => {
    const skill = await client.getSkill(name);
    const content: string = (skill as any).content ?? (skill as any).spec?.content ?? "";
    expect(content).toContain("second");
  });
});

// ---------------------------------------------------------------------------
// Secrets CRUD
// ---------------------------------------------------------------------------

describe("Secrets CRUD", () => {
  const name = "sdk-test-ts-secret-crud";

  beforeAll(async () => {
    await tryDelete(() => client.deleteSecret(name));
  });

  afterAll(async () => {
    await tryDelete(() => client.deleteSecret(name));
  });

  it("creates a secret", async () => {
    const secret = await client.createSecret({
      name,
      data: { MY_KEY: "my-value" },
    });
    expect(secret).toBeDefined();
    expect((secret as any).name).toBe(name);
  });

  it("lists secrets and finds the created one", async () => {
    const list = await client.listSecrets();
    const items: any[] = Array.isArray(list) ? list : (list as any).secrets ?? (list as any).items ?? [];
    const found = items.some((s: any) => s.name === name);
    expect(found).toBe(true);
  });

  it("updates the secret", async () => {
    const updated = await client.updateSecret({
      name,
      data: { MY_KEY: "updated-value" },
    });
    expect(updated).toBeDefined();
  });

  it("deletes the secret", async () => {
    await expect(client.deleteSecret(name)).resolves.not.toThrow();
  });
});

// ---------------------------------------------------------------------------
// Secrets idempotent create
// ---------------------------------------------------------------------------

describe("Secrets idempotent create", () => {
  const name = "sdk-test-ts-secret-idempotent";

  beforeAll(async () => {
    await tryDelete(() => client.deleteSecret(name));
  });

  afterAll(async () => {
    await tryDelete(() => client.deleteSecret(name));
  });

  it("creates a secret the first time", async () => {
    await expect(
      client.createSecret({ name, data: { TOKEN: "first-token" } })
    ).resolves.not.toThrow();
  });

  it("creates the same secret again with different data without error", async () => {
    await expect(
      client.createSecret({ name, data: { TOKEN: "second-token" } })
    ).resolves.not.toThrow();
  });

  it("lists secrets and still finds the entry", async () => {
    const list = await client.listSecrets();
    const items: any[] = Array.isArray(list) ? list : (list as any).secrets ?? (list as any).items ?? [];
    const found = items.some((s: any) => s.name === name);
    expect(found).toBe(true);
  });
});

File: komputer-sdk/go/integration_test.go
package client

import (
	"context"
	"errors"
	"net/http"
	"os"
	"testing"
	"time"
)

// skipIfNoAPI skips the test if the komputer API is not reachable.
// Set KOMPUTER_API_URL to override the default localhost:8080.
func skipIfNoAPI(t *testing.T) string {
	t.Helper()
	url := os.Getenv("KOMPUTER_API_URL")
	if url == "" {
		url = "http://localhost:8080"
		resp, err := http.Get(url + "/api/v1/agents")
		if err != nil {
			t.Skip("Skipping integration tests: API not available at " + url)
		}
		resp.Body.Close()
	}
	return url
}

func newTestClient(t *testing.T) (*Client, context.Context) {
	t.Helper()
	url := skipIfNoAPI(t)
	return New(url), context.Background()
}

// --- Agents E2E ---

func TestIntegration_AgentE2E(t *testing.T) {
	c, ctx := newTestClient(t)
	name := "sdk-test-go-e2e"

	// Clean up before and after — best effort.
	c.DeleteAgent(ctx, name) //nolint:errcheck
	defer c.DeleteAgent(context.Background(), name) //nolint:errcheck

	// Create the agent without a lifecycle opt so it actually runs.
	_, _, err := c.CreateAgent(ctx, name, "Reply with exactly: hello sdk",
		CreateAgentOpts{Model: PtrString("claude-sonnet-4-6")})
	if err != nil {
		t.Fatalf("CreateAgent: %v", err)
	}

	// Watch the agent — prefetches history and opens a live WebSocket.
	stream, err := c.WatchAgent(ctx, name)
	if err != nil {
		t.Fatalf("WatchAgent: %v", err)
	}
	defer stream.Close()

	// Collect events until task_completed or timeout.
	watchCtx, cancel := context.WithTimeout(ctx, 120*time.Second)
	defer cancel()

	var (
		gotText      bool
		gotCompleted bool
	)

	for {
		// Honour context cancellation between reads.
		select {
		case <-watchCtx.Done():
			t.Fatalf("timed out waiting for task_completed (gotText=%v)", gotText)
		default:
		}

		event, err := stream.Next()
		if err != nil {
			// A context cancellation surfaces as a websocket close error.
			if errors.Is(err, context.DeadlineExceeded) || errors.Is(err, context.Canceled) {
				t.Fatalf("stream cancelled before task_completed (gotText=%v): %v", gotText, err)
			}
			// Any other read error after we already got task_completed is fine.
			if gotCompleted {
				break
			}
			t.Fatalf("stream.Next: %v", err)
		}

		switch event.Type {
		case "text":
			gotText = true
		case "task_completed":
			gotCompleted = true
		}

		if gotCompleted {
			break
		}
	}

	if !gotText {
		t.Error("expected at least one event with Type == \"text\", got none")
	}
	if !gotCompleted {
		t.Error("expected an event with Type == \"task_completed\", got none")
	}
}

// --- Memories ---

func TestIntegration_MemoryCRUD(t *testing.T) {
	c, ctx := newTestClient(t)
	name := "sdk-test-go-mem-crud"

	// cleanup before and after
	c.DeleteMemory(ctx, name) //nolint:errcheck
	t.Cleanup(func() { c.DeleteMemory(ctx, name) }) //nolint:errcheck

	// Create
	created, _, err := c.CreateMemory(ctx, name, "initial content",
		CreateMemoryOpts{Description: PtrString("test memory description")})
	if err != nil {
		t.Fatalf("CreateMemory: %v", err)
	}
	if created.GetName() != name {
		t.Errorf("expected name %q, got %q", name, created.GetName())
	}
	if created.GetContent() != "initial content" {
		t.Errorf("expected content %q, got %q", "initial content", created.GetContent())
	}
	if created.GetDescription() != "test memory description" {
		t.Errorf("expected description %q, got %q", "test memory description", created.GetDescription())
	}

	// Get
	got, _, err := c.GetMemory(ctx, name)
	if err != nil {
		t.Fatalf("GetMemory: %v", err)
	}
	if got.GetName() != name {
		t.Errorf("Get: expected name %q, got %q", name, got.GetName())
	}
	if got.GetContent() != "initial content" {
		t.Errorf("Get: expected content %q, got %q", "initial content", got.GetContent())
	}

	// List — verify the created memory appears
	list, _, err := c.ListMemories(ctx)
	if err != nil {
		t.Fatalf("ListMemories: %v", err)
	}
	if !memoryListContains(list, name) {
		t.Errorf("ListMemories: expected %q to appear in list", name)
	}

	// Patch
	patched, _, err := c.PatchMemory(ctx, name, PatchMemoryOpts{Content: PtrString("patched content")})
	if err != nil {
		t.Fatalf("PatchMemory: %v", err)
	}
	if patched.GetContent() != "patched content" {
		t.Errorf("Patch: expected content %q, got %q", "patched content", patched.GetContent())
	}

	// Delete
	_, _, err = c.DeleteMemory(ctx, name)
	if err != nil {
		t.Fatalf("DeleteMemory: %v", err)
	}

	// Verify deleted — GetMemory should now fail
	_, httpResp, err := c.GetMemory(ctx, name)
	if err == nil {
		t.Error("expected error after delete, got nil")
	}
	if httpResp != nil && httpResp.StatusCode != http.StatusNotFound {
		t.Errorf("expected 404 after delete, got %d", httpResp.StatusCode)
	}
}

func TestIntegration_MemoryIdempotentCreate(t *testing.T) {
	c, ctx := newTestClient(t)
	name := "sdk-test-go-mem-idem"

	c.DeleteMemory(ctx, name) //nolint:errcheck
	t.Cleanup(func() { c.DeleteMemory(ctx, name) }) //nolint:errcheck

	// First create
	_, _, err := c.CreateMemory(ctx, name, "original content")
	if err != nil {
		t.Fatalf("first CreateMemory: %v", err)
	}

	// Second create with different content — must not error (idempotent upsert)
	_, _, err = c.CreateMemory(ctx, name, "updated content")
	if err != nil {
		t.Fatalf("idempotent CreateMemory: %v", err)
	}

	// Verify content was updated
	got, _, err := c.GetMemory(ctx, name)
	if err != nil {
		t.Fatalf("GetMemory after idempotent create: %v", err)
	}
	if got.GetContent() != "updated content" {
		t.Errorf("expected content %q after idempotent create, got %q", "updated content", got.GetContent())
	}
}

// --- Skills ---

func TestIntegration_SkillCRUD(t *testing.T) {
	c, ctx := newTestClient(t)
	name := "sdk-test-go-skill-crud"

	c.DeleteSkill(ctx, name) //nolint:errcheck
	t.Cleanup(func() { c.DeleteSkill(ctx, name) }) //nolint:errcheck

	// Create
	created, _, err := c.CreateSkill(ctx, name, "echo hello", "prints hello")
	if err != nil {
		t.Fatalf("CreateSkill: %v", err)
	}
	if created.GetName() != name {
		t.Errorf("expected name %q, got %q", name, created.GetName())
	}
	if created.GetContent() != "echo hello" {
		t.Errorf("expected content %q, got %q", "echo hello", created.GetContent())
	}
	if created.GetDescription() != "prints hello" {
		t.Errorf("expected description %q, got %q", "prints hello", created.GetDescription())
	}

	// Get
	got, _, err := c.GetSkill(ctx, name)
	if err != nil {
		t.Fatalf("GetSkill: %v", err)
	}
	if got.GetName() != name {
		t.Errorf("Get: expected name %q, got %q", name, got.GetName())
	}
	if got.GetContent() != "echo hello" {
		t.Errorf("Get: expected content %q, got %q", "echo hello", got.GetContent())
	}

	// List — verify skill appears
	list, _, err := c.ListSkills(ctx)
	if err != nil {
		t.Fatalf("ListSkills: %v", err)
	}
	if !skillListContains(list, name) {
		t.Errorf("ListSkills: expected %q to appear in list", name)
	}

	// Patch
	patched, _, err := c.PatchSkill(ctx, name, PatchSkillOpts{
		Content:     PtrString("echo world"),
		Description: PtrString("prints world"),
	})
	if err != nil {
		t.Fatalf("PatchSkill: %v", err)
	}
	if patched.GetContent() != "echo world" {
		t.Errorf("Patch: expected content %q, got %q", "echo world", patched.GetContent())
	}
	if patched.GetDescription() != "prints world" {
		t.Errorf("Patch: expected description %q, got %q", "prints world", patched.GetDescription())
	}

	// Delete
	_, _, err = c.DeleteSkill(ctx, name)
	if err != nil {
		t.Fatalf("DeleteSkill: %v", err)
	}

	// Verify deleted
	_, httpResp, err := c.GetSkill(ctx, name)
	if err == nil {
		t.Error("expected error after delete, got nil")
	}
	if httpResp != nil && httpResp.StatusCode != http.StatusNotFound {
		t.Errorf("expected 404 after delete, got %d", httpResp.StatusCode)
	}
}

func TestIntegration_SkillIdempotentCreate(t *testing.T) {
	c, ctx := newTestClient(t)
	name := "sdk-test-go-skill-idem"

	c.DeleteSkill(ctx, name) //nolint:errcheck
	t.Cleanup(func() { c.DeleteSkill(ctx, name) }) //nolint:errcheck

	// First create
	_, _, err := c.CreateSkill(ctx, name, "original content", "original description")
	if err != nil {
		t.Fatalf("first CreateSkill: %v", err)
	}

	// Second create with different content — must not error
	_, _, err = c.CreateSkill(ctx, name, "updated content", "updated description")
	if err != nil {
		t.Fatalf("idempotent CreateSkill: %v", err)
	}

	// Verify content was updated
	got, _, err := c.GetSkill(ctx, name)
	if err != nil {
		t.Fatalf("GetSkill after idempotent create: %v", err)
	}
	if got.GetContent() != "updated content" {
		t.Errorf("expected content %q after idempotent create, got %q", "updated content", got.GetContent())
	}
	if got.GetDescription() != "updated description" {
		t.Errorf("expected description %q after idempotent create, got %q", "updated description", got.GetDescription())
	}
}

// --- Secrets ---

func TestIntegration_SecretCRUD(t *testing.T) {
	c, ctx := newTestClient(t)
	name := "sdk-test-go-secret-crud"

	c.DeleteSecret(ctx, name) //nolint:errcheck
	t.Cleanup(func() { c.DeleteSecret(ctx, name) }) //nolint:errcheck

	// Create
	created, _, err := c.CreateSecret(ctx, name, map[string]string{"key1": "value1", "key2": "value2"})
	if err != nil {
		t.Fatalf("CreateSecret: %v", err)
	}
	if created.GetName() != name {
		t.Errorf("expected name %q, got %q", name, created.GetName())
	}
	// The API returns key names (not values) — verify both keys are present
	keys := created.GetKeys()
	if !containsString(keys, "key1") || !containsString(keys, "key2") {
		t.Errorf("expected keys [key1 key2], got %v", keys)
	}

	// List — verify the secret appears
	list, _, err := c.ListSecrets(ctx)
	if err != nil {
		t.Fatalf("ListSecrets: %v", err)
	}
	found := false
	for _, s := range list.GetSecrets() {
		if s.GetName() == name {
			found = true
			break
		}
	}
	if !found {
		t.Errorf("ListSecrets: expected %q to appear in list", name)
	}

	// Update
	_, _, err = c.UpdateSecret(ctx, name, map[string]string{"key3": "value3"})
	if err != nil {
		t.Fatalf("UpdateSecret: %v", err)
	}

	// Delete
	_, _, err = c.DeleteSecret(ctx, name)
	if err != nil {
		t.Fatalf("DeleteSecret: %v", err)
	}
}

func TestIntegration_SecretIdempotentCreate(t *testing.T) {
	c, ctx := newTestClient(t)
	name := "sdk-test-go-secret-idem"

	c.DeleteSecret(ctx, name) //nolint:errcheck
	t.Cleanup(func() { c.DeleteSecret(ctx, name) }) //nolint:errcheck

	// First create
	_, _, err := c.CreateSecret(ctx, name, map[string]string{"foo": "bar"})
	if err != nil {
		t.Fatalf("first CreateSecret: %v", err)
	}

	// Second create with different data — must not error (client upserts on 409)
	_, _, err = c.CreateSecret(ctx, name, map[string]string{"baz": "qux"})
	if err != nil {
		t.Fatalf("idempotent CreateSecret: %v", err)
	}
}

// --- helpers ---

// memoryListContains checks whether a memory named `name` exists in the raw
// list response (map[string]interface{} keyed by "memories").
func memoryListContains(list map[string]interface{}, name string) bool {
	return namedItemExistsInList(list, "memories", name)
}

// skillListContains checks whether a skill named `name` exists in the raw
// list response (map[string]interface{} keyed by "skills").
func skillListContains(list map[string]interface{}, name string) bool {
	return namedItemExistsInList(list, "skills", name)
}

// namedItemExistsInList searches a list response map for an item with the given name.
// The list response is expected to have a top-level key (e.g. "memories" or "skills")
// whose value is []interface{}, each element being a map[string]interface{} with a
// "name" key.
func namedItemExistsInList(list map[string]interface{}, key string, name string) bool {
	raw, ok := list[key]
	if !ok {
		return false
	}
	items, ok := raw.([]interface{})
	if !ok {
		return false
	}
	for _, item := range items {
		m, ok := item.(map[string]interface{})
		if !ok {
			continue
		}
		if n, _ := m["name"].(string); n == name {
			return true
		}
	}
	return false
}

// containsString reports whether slice contains s.
func containsString(slice []string, s string) bool {
	for _, v := range slice {
		if v == s {
			return true
		}
	}
	return false
}
